### PR TITLE
ftr(etcd): add benchmarks for getlist

### PR DESCRIPTION
**Pull request by sxllwx**
#### What this PR does / why we need it:

ftr(etcd): add benchmarks for getlist.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 1]
```diff
@@ -622,3 +528,68 @@ func TestInvalidKeys(t *testing.T) {
...
+			ctx, store, etcdClient := testSetup(b)
+			defer etcdClient.Close()
+
+			// make fake pods..
+			dir := "/testing"
+			key := dir + "/testkey"
+			originalRevision := ""
+			for i := 0; i < tc.objectNum; i++ {
+				out := &example.Pod{}
+				obj := &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("pod-%d", i), SelfLink: "testlink"}}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:573]: Hmm - this is fairly small object.
Can we try having some more-representative object?

TBH, I'm a bit surprised that allocating a new slice (even though it's just pointers) is actually universally faster...

Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:573]: > Can we try having some more-representative object?

I checked, examplev1.Pod is indeed relatively small. What do you think of corev1.Pod and corev1.Node?

**Comment by sxllwx** [comment]
Update: 

````bash
# benchcmp 
benchmark                                            old ns/op      new ns/op      delta
BenchmarkStore_GetList/list_10_corev1.Pod-16         475431         453929         -4.52%
BenchmarkStore_GetList/list_10_corev1.Pod-16         464205         448608         -3.36%
BenchmarkStore_GetList/list_100_corev1.Pod-16        2815236        2575968        -8.50%
BenchmarkStore_GetList/list_100_corev1.Pod-16        2913277        2569991        -11.78%
BenchmarkStore_GetList/list_1000_corev1.Pod-16       26287639       24295101       -7.58%
BenchmarkStore_GetList/list_1000_corev1.Pod-16       26641453       23947811       -10.11%
BenchmarkStore_GetList/list_10000_corev1.Pod-16      252053668      238738248      -5.28%
BenchmarkStore_GetList/list_10000_corev1.Pod-16      272613894      234570819      -13.95%
BenchmarkStore_GetList/list_50000_corev1.Pod-16      1213607441     1214340396     +0.06%
BenchmarkStore_GetList/list_50000_corev1.Pod-16      1194805913     1197901030     +0.26%
BenchmarkStore_GetList/list_10_corev1.Node-16        477939         454682         -4.87%
BenchmarkStore_GetList/list_10_corev1.Node-16        484702         453219         -6.50%
BenchmarkStore_GetList/list_100_corev1.Node-16       2767087        2656915        -3.98%
BenchmarkStore_GetList/list_100_corev1.Node-16       2770697        2641017        -4.68%
BenchmarkStore_GetList/list_1000_corev1.Node-16      26196409       24679840       -5.79%
BenchmarkStore_GetList/list_1000_corev1.Node-16      26459178       24623285       -6.94%
BenchmarkStore_GetList/list_10000_corev1.Node-16     261074212      242936736      -6.95%
BenchmarkStore_GetList/list_10000_corev1.Node-16     254321802      243070395      -4.42%
BenchmarkStore_GetList/list_50000_corev1.Node-16     1231968582     1215763255     -1.32%
BenchmarkStore_GetList/list_50000_corev1.Node-16     1209250275     1207361772     -0.16%

benchmark                                            old allocs     new allocs     delta
BenchmarkStore_GetList/list_10_corev1.Pod-16         900            898            -0.22%
BenchmarkStore_GetList/list_10_corev1.Pod-16         887            902            +1.69%
BenchmarkStore_GetList/list_100_corev1.Pod-16        6119           5953           -2.71%
BenchmarkStore_GetList/list_100_corev1.Pod-16        6040           6033           -0.12%
BenchmarkStore_GetList/list_1000_corev1.Pod-16       60508          59499          -1.67%
BenchmarkStore_GetList/list_1000_corev1.Pod-16       60529          59523          -1.66%
BenchmarkStore_GetList/list_10000_corev1.Pod-16      609795         600322         -1.55%
BenchmarkStore_GetList/list_10000_corev1.Pod-16      610197         600522         -1.59%
BenchmarkStore_GetList/list_50000_corev1.Pod-16      3052665        3002197        -1.65%
BenchmarkStore_GetList/list_50000_corev1.Pod-16      3052297        3002216        -1.64%
BenchmarkStore_GetList/list_10_corev1.Node-16        955            934            -2.20%
BenchmarkStore_GetList/list_10_corev1.Node-16        937            934            -0.32%
BenchmarkStore_GetList/list_100_corev1.Node-16       6541           6455           -1.31%
BenchmarkStore_GetList/list_100_corev1.Node-16       6541           6453           -1.35%
BenchmarkStore_GetList/list_1000_corev1.Node-16      65397          64456          -1.44%
BenchmarkStore_GetList/list_1000_corev1.Node-16      65748          64562          -1.80%
BenchmarkStore_GetList/list_10000_corev1.Node-16     660805         649795         -1.67%
BenchmarkStore_GetList/list_10000_corev1.Node-16     659782         649671         -1.53%
BenchmarkStore_GetList/list_50000_corev1.Node-16     3302058        3252525        -1.50%
BenchmarkStore_GetList/list_50000_corev1.Node-16     3302582        3251713        -1.54%

benchmark                                            old bytes     new bytes     delta
BenchmarkStore_GetList/list_10_corev1.Pod-16         143617        105911        -26.25%
BenchmarkStore_GetList/list_10_corev1.Pod-16         143836        105992        -26.31%
BenchmarkStore_GetList/list_100_corev1.Pod-16        1271534       877053        -31.02%
BenchmarkStore_GetList/list_100_corev1.Pod-16        1382780       879686        -36.38%
BenchmarkStore_GetList/list_1000_corev1.Pod-16       12796879      8759114       -31.55%
BenchmarkStore_GetList/list_1000_corev1.Pod-16       12792836      8747185       -31.62%
BenchmarkStore_GetList/list_10000_corev1.Pod-16      118933402     88056604      -25.96%
BenchmarkStore_GetList/list_10000_corev1.Pod-16      118966072     88100177      -25.95%
BenchmarkStore_GetList/list_50000_corev1.Pod-16      441080400     444088304     +0.68%
BenchmarkStore_GetList/list_50000_corev1.Pod-16      441022056     444101616     +0.70%
BenchmarkStore_GetList/list_10_corev1.Node-16        109802        85715         -21.94%
BenchmarkStore_GetList/list_10_corev1.Node-16        117276        85604         -27.01%
BenchmarkStore_GetList/list_100_corev1.Node-16       927463        676136        -27.10%
BenchmarkStore_GetList/list_100_corev1.Node-16       931019        674296        -27.57%
BenchmarkStore_GetList/list_1000_corev1.Node-16      9474277       6661856       -29.68%
BenchmarkStore_GetList/list_1000_corev1.Node-16      9495070       6647492       -29.99%
BenchmarkStore_GetList/list_10000_corev1.Node-16     87340370      67068329      -23.21%
BenchmarkStore_GetList/list_10000_corev1.Node-16     87347420      67049544      -23.24%
BenchmarkStore_GetList/list_50000_corev1.Node-16     333308272     336621728     +0.99%
BenchmarkStore_GetList/list_50000_corev1.Node-16     333246248     336292280     +0.91%
````

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 1 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: Fixed,  PTAL @wojtek-t

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 2]
```diff
@@ -30,7 +29,9 @@ import (
 	"go.etcd.io/etcd/server/v3/embed"
 	"google.golang.org/grpc/grpclog"
 
+	corev1 "k8s.io/api/core/v1"
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:32]: Let's not add this - we can create a non-empty object with examplev1, so let's just use that only.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 3]
```diff
@@ -622,3 +531,163 @@ func TestInvalidKeys(t *testing.T) {
...
+							Running: &corev1.ContainerStateRunning{
+								StartedAt: metav1.Time{},
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+	generateNode := func() runtime.Object {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:569]: I don't think having node here actually provides much value. So let's remove it.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 4]
```diff
@@ -622,3 +531,163 @@ func TestInvalidKeys(t *testing.T) {
 	_, countErr := store.Count(invalidKey)
 	expectInvalidKey("Count", countErr)
 }
+
+func BenchmarkStore_GetList(b *testing.B) {
+	generatePod := func() runtime.Object {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:536]: This object is still fairly small in data.

I would like to see something like "generateBigPod" or sth (and you can approximate it by putting large annotations on it).

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 5]
```diff
@@ -622,3 +531,163 @@ func TestInvalidKeys(t *testing.T) {
...
+
+			// make fake objects..
+			dir := "/testing"
+			key := dir + "/testkey"
+			originalRevision := ""
+			for i := 0; i < tc.objectNum; i++ {
+				obj := tc.newObjectFunc()
+				o := obj.(metav1.Object)
+				o.SetName(fmt.Sprintf("object-%d", i))
+				o.SetSelfLink("testlink")
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:670]: Let's not set SelfLink - selflinks are deprecated

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 6]
```diff
@@ -622,3 +531,163 @@ func TestInvalidKeys(t *testing.T) {
...
+	}
+	for _, tc := range testCases {
+		b.Run(tc.name, func(b *testing.B) {
+			// booting etcd instance
+			ctx, store, etcdClient := testSetup(b)
+			defer etcdClient.Close()
+
+			// make fake objects..
+			dir := "/testing"
+			key := dir + "/testkey"
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:664]: seems unused

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 7]
```diff
@@ -622,3 +531,163 @@ func TestInvalidKeys(t *testing.T) {
...
+			list := tc.newListObjectFunc()
+
+			// now we start benchmarking
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				if err := store.GetList(ctx, dir, storage.ListOptions{Predicate: storage.Everything, Recursive: true}, list); err != nil {
+					b.Errorf("Unexpected List error: %v", err)
+				}
+				finalRevision := list.(metav1.ListInterface).GetResourceVersion()
+				if originalRevision != finalRevision {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:687]: Let's verify if the number of objects returned is the expected.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go [hunk 8]
```diff
@@ -622,3 +531,163 @@ func TestInvalidKeys(t *testing.T) {
...
+					b.Fatalf("Set failed: %v", err)
+				}
+				originalRevision = out.(metav1.Object).GetResourceVersion()
+			}
+			list := tc.newListObjectFunc()
+
+			// now we start benchmarking
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				if err := store.GetList(ctx, dir, storage.ListOptions{Predicate: storage.Everything, Recursive: true}, list); err != nil {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:683]: This is still returning all objects.
I wanted to see a case where we return only a relatively small subset of objects (say 1% or sth like that).

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 6 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 5 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 2 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 3 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 4 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 7 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go (hunk 8 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go]: fixed, PTAL @wojtek-t

**Comment by sxllwx** [comment]
```bash
benchmark                                                        old ns/op      new ns/op      delta
BenchmarkStore_GetList/pick_1000_pods_out_of_100000_pod-16       6269308270     6250230597     -0.30%
BenchmarkStore_GetList/pick_1000_pods_out_of_100000_pod-16       6287190432     6266327522     -0.33%
BenchmarkStore_GetList/pick_10000_Pods_out_of_100000_pod-16      6299670296     6303797752     +0.07%
BenchmarkStore_GetList/pick_10000_Pods_out_of_100000_pod-16      6317763922     6305945383     -0.19%
BenchmarkStore_GetList/pick_25000_Pods_out_of_100000_pod-16      6312331931     6352382247     +0.63%
BenchmarkStore_GetList/pick_25000_Pods_out_of_100000_pod-16      6352589171     6387865500     +0.56%
BenchmarkStore_GetList/pick_50000_Pods_out_of_100000_pod-16      6387484883     6376653645     -0.17%
BenchmarkStore_GetList/pick_50000_Pods_out_of_100000_pod-16      6426496866     6362955333     -0.99%
BenchmarkStore_GetList/pick_100000_Pods_out_of_100000_pod-16     6556018553     6428002054     -1.95%
BenchmarkStore_GetList/pick_100000_Pods_out_of_100000_pod-16     6544365720     6437920403     -1.63%

benchmark                                                        old allocs     new allocs     delta
BenchmarkStore_GetList/pick_1000_pods_out_of_100000_pod-16       5124489        5124203        -0.01%
BenchmarkStore_GetList/pick_1000_pods_out_of_100000_pod-16       5122806        5118873        -0.08%
BenchmarkStore_GetList/pick_10000_Pods_out_of_100000_pod-16      5202060        5200043        -0.04%
BenchmarkStore_GetList/pick_10000_Pods_out_of_100000_pod-16      5202554        5192521        -0.19%
BenchmarkStore_GetList/pick_25000_Pods_out_of_100000_pod-16      5338729        5310609        -0.53%
BenchmarkStore_GetList/pick_25000_Pods_out_of_100000_pod-16      5336929        5314654        -0.42%
BenchmarkStore_GetList/pick_50000_Pods_out_of_100000_pod-16      5564591        5513459        -0.92%
BenchmarkStore_GetList/pick_50000_Pods_out_of_100000_pod-16      5562133        5512893        -0.89%
BenchmarkStore_GetList/pick_100000_Pods_out_of_100000_pod-16     6015314        5912202        -1.71%
BenchmarkStore_GetList/pick_100000_Pods_out_of_100000_pod-16     6011989        5912739        -1.65%

benchmark                                                        old bytes      new bytes      delta
BenchmarkStore_GetList/pick_1000_pods_out_of_100000_pod-16       1518817664     1518214144     -0.04%
BenchmarkStore_GetList/pick_1000_pods_out_of_100000_pod-16       1518802912     1517903264     -0.06%
BenchmarkStore_GetList/pick_10000_Pods_out_of_100000_pod-16      1541939920     1528406264     -0.88%
BenchmarkStore_GetList/pick_10000_Pods_out_of_100000_pod-16      1542009352     1528227720     -0.89%
BenchmarkStore_GetList/pick_25000_Pods_out_of_100000_pod-16      1588886120     1545267600     -2.75%
BenchmarkStore_GetList/pick_25000_Pods_out_of_100000_pod-16      1588706968     1545447152     -2.72%
BenchmarkStore_GetList/pick_50000_Pods_out_of_100000_pod-16      1666975368     1573613744     -5.60%
BenchmarkStore_GetList/pick_50000_Pods_out_of_100000_pod-16      1666638248     1573729472     -5.57%
BenchmarkStore_GetList/pick_100000_Pods_out_of_100000_pod-16     1885756232     1630320928     -13.55%
BenchmarkStore_GetList/pick_100000_Pods_out_of_100000_pod-16     1885888072     1630109920     -13.56%
```

**Comment by stevekuznetsov** [comment]
Is [this](https://github.com/kubernetes/kubernetes/pull/116374#issuecomment-1466421479) the most up-to-date benchmark? So this improves performance by ~1%?

**Comment by sxllwx** [comment]
> Is [this](https://github.com/kubernetes/kubernetes/pull/116374#issuecomment-1466421479) the most up-to-date benchmark? So this improves performance by ~1%?

Yes, but I don't think the results of this benchmark seem convincing.

This PR avoids repeated calls to reflect.Append + reflect.Set  (large amount of memory is not actually used). This code is like: 

```golang
podList1 := []Pod{pod-0, pod-1, pod-2, ... pod-1023}
podList2 := append(podList1, pod-1024 .... pod-2048)
....
final := append(podListX, pod-yyy .... pod-xxx)
```

> In fact, the slices generated the first two times will be dropped by GC. Because we only need the final `final` slice


Because the size of each exmaplev1.Pod itself is relatively small (520byte) `fmt.Println(unsafe.Sizeof(examplev1.Pod{}))`, compared with the 2K memory requested in the benchmark, it is much smaller. So it seems that the improvement is very small.


The same optimization method can achieve 80% improvement in PR#116327. see https://github.com/kubernetes/kubernetes/pull/116327#issuecomment-1459733316

**Comment by sxllwx** [comment]
@wojtek-t PTAL, thx..

**Comment by MadhavJivrajani** [comment]
This should also fix https://github.com/kubernetes/kubernetes/issues/110734

**Comment by sxllwx** [comment]
> Given the result in benchmarks I don't think it's actually worth pursuing this PR - every change is introducing some additional risk, so given negligible gain, I would personally suggesting closing this PR (or alternatively splitting it to addition of benchmarks and only commit the benchmarks without the actual code changes).

thank you for your reply.

All right. I will add benchmark and modify getlist into two PRs to submit. I have changed the current PR to just add the benchmark (PTAL thx), and I will resubmit a PR to change the getlist later.

**Pull request merged**